service/history/queue/processing_queue.go

```diff
@@ -0,0 +1,366 @@
+	return q.state
+}
+
+func (q *processingQueueImpl) Split(
+	policy ProcessingQueueSplitPolicy,
+) []ProcessingQueue {
+	newQueueStates := policy.Evaluate(q)
+	if len(newQueueStates) == 0 {
+		// no need to split, return self
+		return []ProcessingQueue{q}
+	}
+
+	return splitProcessingQueue([]*processingQueueImpl{q}, newQueueStates, q.logger, q.metricsClient)
+}
+
+func (q *processingQueueImpl) Merge(
+	queue ProcessingQueue,
+) []ProcessingQueue {
+	q1, q2 := q, queue.(*processingQueueImpl)
+
+	if q1.State().Level() != q2.State().Level() {
+		errMsg := "Processing queue encountered a queue from different level during merge"
+		q.logger.Error(errMsg, tag.Error(
+			fmt.Errorf("current queue level: %v, incoming queue level: %v", q1.state.level, q2.state.level),
+		))
+		panic(errMsg)
+	}
+
+	if !q1.state.ackLevel.Less(q2.state.maxLevel) ||
+		!q2.state.ackLevel.Less(q1.state.maxLevel) {
+		// one queue's ackLevel is larger or equal than the other one's maxLevel
+		// this means there's no overlap between two queues
+		return []ProcessingQueue{q1, q2}
+	}
+
+	// generate new queue states for merged queues
+	newQueueStates := []ProcessingQueueState{}
+	if !taskKeyEquals(q1.state.ackLevel, q2.state.ackLevel) {
+		if q2.state.ackLevel.Less(q1.state.ackLevel) {
+			q1, q2 = q2, q1
+		}
+
+		newQueueStates = append(newQueueStates, newProcessingQueueState(
+			q1.state.level,
+			q1.state.ackLevel,
+			minTaskKey(q1.state.readLevel, q2.state.ackLevel),
+			q2.state.ackLevel,
+			q1.state.domainFilter.copy(),
+		))
+	}
+
+	if !taskKeyEquals(q1.state.maxLevel, q2.state.maxLevel) {
+		if q1.state.maxLevel.Less(q2.state.maxLevel) {
+			q1, q2 = q2, q1
+		}
+
+		newQueueStates = append(newQueueStates, newProcessingQueueState(
+			q1.state.level,
+			q2.state.maxLevel,
+			maxTaskKey(q1.state.readLevel, q2.state.maxLevel),
+			q1.state.maxLevel,
+			q1.state.domainFilter.copy(),
+		))
+	}
+
+	newQueueStates = append(newQueueStates, newProcessingQueueState(
+		q1.state.level,
+		maxTaskKey(q1.state.ackLevel, q2.state.ackLevel),
+		minTaskKey(q1.state.readLevel, q2.state.readLevel),
+		minTaskKey(q1.state.maxLevel, q2.state.maxLevel),
+		q1.state.domainFilter.Merge(q2.state.domainFilter),
+	))
+
+	return splitProcessingQueue([]*processingQueueImpl{q1, q2}, newQueueStates, q.logger, q.metricsClient)
+}
+
+func (q *processingQueueImpl) AddTasks(
+	tasks map[task.Key]task.Task,
+) {
+	for key, task := range tasks {
+		if _, loaded := q.outstandingTasks[key]; loaded {
+			q.logger.Debug(fmt.Sprintf("Skipping task: %+v. DomainID: %v, WorkflowID: %v, RunID: %v, Type: %v",
+				key, task.GetDomainID(), task.GetWorkflowID(), task.GetRunID(), task.GetTaskType()))
+			continue
+		}
+
+		if !taskBelongsToProcessQueue(q.state, key, task) {
+			errMsg := "Processing queue encountered a task doesn't belong to its scope"
+			q.logger.Error(errMsg, tag.Error(
+				fmt.Errorf("Processing queue state: %+v, task: %+v", q.state, key),
+			))
+			panic(errMsg)
```

Comment:
Is this a sanity check and should panic?

Reply:
Yes this is just a sanity check. If this happens, it means there's a bug in the implementation for queue collection. and yes, the code will panic on L250 if this happens.